api/internal/git/repospec.go

```diff
@@ -212,18 +213,26 @@ func peelQuery(arg string) (string, string, time.Duration, bool) {
 	return parsed.Path, ref, duration, submodules
 }
 
+var userRegex = regexp.MustCompile(`^[a-zA-Z][a-zA-Z0-9-]*@`)
+
 func parseHostSpec(n string) (string, string) {
 	var host string
-	// Start accumulating the host part.
-	for _, p := range []string{
-		// Order matters here.
-		"git::", "gh:", "ssh://", "https://", "http://", "file://",
-		"git@", "github.com:", "github.com/"} {
-		if len(p) < len(n) && strings.ToLower(n[:len(p)]) == p {
-			n = n[len(p):]
-			host += p
+	consumeHostStrings := func(parts []string) {
+		for _, p := range parts {
+			if len(p) < len(n) && strings.ToLower(n[:len(p)]) == p {
+				n = n[len(p):]
+				host += p
+			}
 		}
 	}
+	// Start accumulating the host part.
+	// Order matters here.
+	consumeHostStrings([]string{"git::", "gh:", "ssh://", "https://", "http://", "file://"})
+	if p := userRegex.FindString(n); p != "" {
```

Comment:
Since we end up forever supporting everything that we make work, I'm trying to wrap my head around everything this makes possible in addition to what if anything it could break. 
- Since we're relying on the `@` in effect, what other valid paths might contain it, and does our parsing of them change before/after? For example, `https://username:password@github.com/username/repository.git` seems to parse currently, albeit discarding the username/password parts.
- For the purposes of the feature you need, we only need to add support for arbitrary usernames when SSH in particular is being used with Github specifically, but will this make anything else new work? It seems to me arbitrary protocol prefixes will be accepted with arbitrary usernames, but only for Github. Other hosts will continue to only work with the Git user. Should we further restrict this? Any new combos that would work should get test coverage.